sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReadQuerySource.java

```diff
@@ -0,0 +1,179 @@
+      BigQueryServices bqServices,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      Coder<T> coder,
+      ReadSessionOptions readSessionOptions,
+      String stepUuid,
+      ValueProvider<String> query,
+      Boolean flattenResults,
+      Boolean useLegacySql) {
+    super(bqServices, parseFn, coder, readSessionOptions);
+    this.stepUuid = checkNotNull(stepUuid, "stepUuid");
+    this.query = checkNotNull(query, "query");
+    this.flattenResults = checkNotNull(flattenResults, "flattenResults");
+    this.useLegacySql = checkNotNull(useLegacySql, "useLegacySql");
+    this.dryRunJobStats = new AtomicReference<>();
+  }
+
+  @Override
+  public long getEstimatedSizeBytes(PipelineOptions options)
+      throws InterruptedException, IOException {
+    BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
+    return dryRunQueryIfNeeded(bqOptions).getQuery().getTotalBytesProcessed();
+  }
+
+  @Override
+  protected TableReference getEffectiveTableReference(BigQueryOptions bqOptions)
+      throws InterruptedException, IOException {
+    // 1. Find the location of the query.
+    String location = null;
+    List<TableReference> referencedTables =
+        dryRunQueryIfNeeded(bqOptions).getQuery().getReferencedTables();
+    DatasetService tableService = bqServices.getDatasetService(bqOptions);
+    if (referencedTables != null && !referencedTables.isEmpty()) {
+      TableReference queryTable = referencedTables.get(0);
+      location = tableService.getTable(queryTable).getLocation();
+    }
+
+    // 2. Create the temporary dataset in the query location.
+    TableReference tableToRead = createTempTableReference(
+        bqOptions.getProject(), createJobIdToken(bqOptions.getJobName(), stepUuid));
+
+    LOG.info("Creating temporary dataset {} for query results", tableToRead.getDatasetId());
+
+    tableService.createDataset(
+        tableToRead.getProjectId(),
+        tableToRead.getDatasetId(),
+        location,
+        "Temporary tables for query results of job " + bqOptions.getJobName(),
+        TimeUnit.DAYS.toMillis(1));
+
+    // 3. Execute the query.
+    String queryJobId = createJobIdToken(bqOptions.getJobName(), stepUuid) + "-query";
+
+    LOG.info("Writing query results to temporary table {} with job {}", tableToRead, queryJobId);
+
+    executeQuery(
+        bqOptions.getProject(),
+        queryJobId,
+        tableToRead,
+        bqServices.getJobService(bqOptions));
+
+    LOG.info("Query job {} completed", queryJobId);
+
+    return tableToRead;
+  }
+
+  private synchronized JobStatistics dryRunQueryIfNeeded(BigQueryOptions bqOptions)
+      throws InterruptedException, IOException {
+    if (dryRunJobStats.get() == null) {
+      JobStatistics jobStats = bqServices.getJobService(bqOptions).dryRunQuery(
+          bqOptions.getProject(), createBasicQueryConfig());
+      boolean result = dryRunJobStats.compareAndSet(null, jobStats);
+      System.out.print("Result was " + result);
```

Comment:
@kmjung , should we use debug log or remove it?

Reply:
This was left over from some debugging -- I've removed it.